kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,149 @@
+        #asWord ( #buf ( 32 -Int lengthBytes(B2), (A /Int (2 ^Int (8 *Int lengthBytes(B2)))) |Int #asWord ( B1 ) ) +Bytes B2 )
+        requires #rangeUInt(256, A) andBool A modInt (2 ^Int (8 *Int lengthBytes(B2))) ==Int 0 andBool lengthBytes(B1 +Bytes B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 2b. |Int of +Bytes, update to be done in right
+    rule [bor-update-to-right]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( B1 +Bytes #buf ( lengthBytes(B2), A |Int #asWord ( B2 ) ) )
+        requires 0 <=Int A andBool A <Int 2 ^Int (8 *Int lengthBytes(B2)) andBool lengthBytes(B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 3a. Update with explicit shift and symbolic slot
+    rule [bor-update-with-shift]:
+      ( SHIFT *Int X ) |Int Y => #asWord ( #buf( 32 -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, Y ) )
+      requires #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (32 -Int ( log2Int(SHIFT) /Int 8 )))
+       andBool 0 <=Int Y andBool Y <Int SHIFT
+       [simplification, concrete(SHIFT), comm, preserves-definedness]
+
+    // 3b. Buffer cropping
+    rule [buf-asWord-crop]:
+      #buf (W1:Int, #asWord(#buf(W2, X:Int) +Bytes B)) => #buf(W1 -Int lengthBytes(B), X) +Bytes B
+      requires 0 <=Int W1 andBool W1 <=Int 32 andBool 0 <=Int W2 andBool
+       andBool lengthBytes(B) <=Int W1 andBool W1 <=Int W2 +Int lengthBytes(B)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int lengthBytes(B)))
+      [simplification]
+
+    // 3c. Splitting the updated buffer into the updated value and the trailing zeros, explicit shift
+    rule [buf-split-l]:
+      #buf ( W, SHIFT *Int X ) => #buf( W -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, 0)
+      requires 0 <=Int W andBool W <=Int 32 andBool #isByteShift(SHIFT)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W -Int ( log2Int(SHIFT) /Int 8)))
+       [simplification, concrete(W, SHIFT), preserves-definedness]
+
+    // 3d. Splitting the updated buffer into the updated value and the trailing zeros, implicit shift
+    rule [bor-split]:
+      X |Int Y => #asWord ( #buf ( 32 -Int #getFirstOneBit(X) /Int 8, X /Int ( 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) ) ) ) +Bytes
+                            #buf ( #getFirstOneBit(X) /Int 8, Y ) )
+      requires #rangeUInt(256, X) andBool 0 <=Int #getFirstOneBit(X)
+       andBool 0 <=Int Y andBool Y <Int 2 ^Int ( 8 *Int ( #getFirstOneBit(X) /Int 8 ) )
+       [simplification, concrete(X), preserves-definedness]
+
+    // 3e. Clearing update leftovers that only have bits higher than the buffer accepts
+    rule [buf-bor-subsume]:
+      #buf ( N, X |Int Y ) => #buf ( N, X )
+      requires 0 <=Int N andBool N <=Int 32
+       andBool 0 <=Int X andBool X <Int 2 ^Int ( 8 *Int N )
+       andBool Y modInt 2 ^Int ( 8 *Int N ) ==Int 0
+       [simplification, concrete(X), preserves-definedness]
```

Comment:
This should _never_ happen. It's an invariant of `#buf(W, X)` that it cannot be built ill-formed, we must always have `0 <=Int X andBool X <Int 2 ^Int (8 *Int W)`. If this is being violated, we need to track down where and fix it up. I suspect the prior rule may be to blame, it could build an ill-formed `#buf(1, 0)`.

Reply:
I know which engagement this was created for, so I'll re-run the proofs without this rule and see what happens.

Reply:
`andBool Y modInt 2 ^Int ( 8 *Int N ) ==Int 0` is a very strange condition to request of a symbolic `Y` to hold.

Reply:
I will remove this lemma and everything needs to pass without it.